HPCC-19949 Persist_replicate.ecl failed on Roxie after "make install".

AttilaVamos opened the pull request:
Update mspec.defaultBaseDir and mspec.defaultReplicateDir if they are
differ from the group default base directory

Signed-off-by: HPCCSmoketest <hpccsmoketest@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually in HPCC system generated with make install.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
1/1. Process of PR-11748, label: HPCC-19949-fix-7.0.0 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 36dff5dac40f055cfb8688d56e94089b37eb5e5a
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   23 sec (00:00:23) |  186 sec (00:03:06) |   95 sec (00:01:35) |   22 sec (00:00:22) |   44 sec (00:00:44) | 1473 sec (00:24:33) |   18 sec (00:00:18) | 1861 sec (00:31:01) |

AttilaVamos commented:
@jakesmith please review.

jakesmith commented on dali/base/dafdesc.cpp:
```diff
@@ -432,11 +432,14 @@ struct CClusterInfo: implements IClusterInfo, public CInterface
+                baseDirChanged = true;
             }
-            if (mspec.defaultCopies>1 && mspec.defaultReplicateDir.isEmpty())
+            if ((mspec.defaultCopies>1 && mspec.defaultReplicateDir.isEmpty()) || baseDirChanged)
```
Shouldn't this be:
```
if (mspec.defaultCopies>1 && (mspec.defaultReplicateDir.isEmpty() || baseDirChanged))
```

i.e. shouldn't be setting replicate dir if <=1 copies ..

AttilaVamos replied:
Changed

jakesmith commented:
@AttilaVamos - please see comments.

AttilaVamos commented on dali/base/dafdesc.cpp:
```diff
@@ -432,11 +432,14 @@ struct CClusterInfo: implements IClusterInfo, public CInterface
-            if (mspec.defaultBaseDir.isEmpty())
+            bool baseDirChanged = false;
+            if (mspec.defaultBaseDir.isEmpty() ||
+               (!defaultDir.isEmpty() && (strcmp(mspec.defaultBaseDir.str(),defaultDir.str()) != 0)) )
```
Changed

AttilaVamos commented:
@jakesmith updated, please check again.

HPCCSmoketest commented:
1/1. Process of PR-11748, label: HPCC-19949-fix-7.0.0 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 77323af5247a07d8b4f3361f7b7cad461d2ac858
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  185 sec (00:03:05) |   95 sec (00:01:35) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1477 sec (00:24:37) |   19 sec (00:00:19) | 1858 sec (00:30:58) |

jakesmith commented:
@AttilaVamos - looks good.

@richardkchapman - ready to merge

richardkchapman commented:
@AttilaVamos Please squash

AttilaVamos committed:
HPCC-19949 Persist_replicate.ecl failed on Roxie after "make install".

Update mspec.defaultBaseDir and mspec.defaultReplicateDir if they are
differ from the group default base directory

Tested manually HPCC system generated with make install.

Signed-off-by: HPCCSmoketest <hpccsmoketest@gmail.com>

AttilaVamos commented:
@richardkchapman done

HPCCSmoketest commented:
This PR (11748) is in the 1/1 position of the queue. Estimated start time is after ~0.5 hour(s)

HPCCSmoketest commented:
2/2. Process of PR-11748, label: HPCC-19949-fix-7.0.0 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0130db8b99143d86579171df2cb96d46b7994d68
Build: success
Install hpccsystems-platform-community_7.0.0-rc2.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |806 |806 |0 |
|test (thor) |730 |730 |0 |
|test (roxie) |879 |879 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  186 sec (00:03:06) |   95 sec (00:01:35) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1499 sec (00:24:59) |   19 sec (00:00:19) | 1880 sec (00:31:20) |
